### PR TITLE
feat: load gqlgen plugins during parse schema

### PR DESCRIPTION
**Pull request by giautm**
Currently, the EntGQL extension didn't load any GQLGen plugins. So, When I use Federations (it's a built-in GQLGen plugin), EntGQL complains about undefined `@key` directives which will be defined by the plugin via code-mod.
```gql
# Todo with Federations enabled
type Todo implements Node @key(fields: "id") {
  id: ID!
  createdAt: Time
  status: Status!
  priority: Int!
  text: String!
  parent: Todo
  children: [Todo!]
}
```

Current error
```
2022/01/05 10:43:17 creating EntGQL extension: internal/graphql/schema/schema.gql:54: Undefined directive key.
exit status 1
```

This PR loads built-in plugins from GQLGen, and also provides options to load third-party plugins.

### CODE REVIEW

File: entgql/extension.go [hunk 1]
```diff
@@ -109,8 +114,44 @@ func WithConfigPath(path string) ExtensionOption {
...
+						cfg.Sources = append(cfg.Sources, s)
+					}
+				}
+			}
+
 			if err := cfg.LoadSchema(); err != nil {
-				return err
+				return fmt.Errorf("failed to load schema: %w", err)
+			}
+
```
Comment (masseelch) [inline entgql/extension.go:143]:
```suggestion
```

File: entgql/extension.go [hunk 2]
```diff
@@ -109,8 +114,44 @@ func WithConfigPath(path string) ExtensionOption {
...
+			for _, p := range plugins {
+				if inj, ok := p.(plugin.LateSourceInjector); ok {
+					if s := inj.InjectSourceLate(cfg.Schema); s != nil {
+						cfg.Sources = append(cfg.Sources, s)
+					}
+				}
+			}
+
```
Comment (masseelch) [inline entgql/extension.go:151]:
```suggestion
```

File: entgql/extension.go [hunk 3]
```diff
@@ -109,8 +114,44 @@ func WithConfigPath(path string) ExtensionOption {
...
+			for _, p := range plugins {
+				if inj, ok := p.(plugin.EarlySourceInjector); ok {
+					if s := inj.InjectSourceEarly(); s != nil {
+						cfg.Sources = append(cfg.Sources, s)
+					}
+				}
+			}
+
```
Comment (masseelch) [inline entgql/extension.go:139]:
```suggestion
```

File: entgql/extension.go [hunk 4]
```diff
@@ -109,8 +114,44 @@ func WithConfigPath(path string) ExtensionOption {
...
+			}
+			plugins = append(plugins, resolvergen.New())
+			if cfg.Federation.IsDefined() {
+				plugins = append([]plugin.Plugin{federation.New()}, plugins...)
+			}
+
+			for _, o := range option {
+				o(cfg, &plugins)
+			}
+
```
Comment (masseelch) [inline entgql/extension.go:131]:
```suggestion
```

File: entgql/extension.go [hunk 5]
```diff
@@ -109,8 +114,44 @@ func WithConfigPath(path string) ExtensionOption {
...
+			// https://github.com/99designs/gqlgen/blob/47015f12e3aa26af251fec67eab50d3388c17efe/api/generate.go#L21-L57
+			plugins := []plugin.Plugin{}
+			if cfg.Model.IsDefined() {
+				plugins = append(plugins, modelgen.New())
+			}
+			plugins = append(plugins, resolvergen.New())
+			if cfg.Federation.IsDefined() {
+				plugins = append([]plugin.Plugin{federation.New()}, plugins...)
+			}
+
```
Comment (masseelch) [inline entgql/extension.go:127]:
```suggestion
```

**Review by masseelch — CHANGES_REQUESTED** [review]
Can you add tests for this?

**Comment by giautm** [comment]
What kind of test? 😁 trying to enable federation?!

**Comment by giautm** [comment]
Hey @masseelch, plz review again. 🥰

**Comment by masseelch** [comment]
This is what I meant yes. Though I advise to let `todo`, `todopulid` and `todouuid` unchanges and simply add another dir `todofed` or similar. Looks great otherwise :+1: 

And we need to see if @a8m is willing to merge it :-)

**Comment by giautm** [comment]
Updated PR with todofed, and unchanges other apps.

**Review by masseelch — APPROVED** [review]
Thank you :-)

File: entgql/extension.go [hunk 6]
```diff
@@ -109,8 +114,39 @@ func WithConfigPath(path string) ExtensionOption {
...
+				return fmt.Errorf("failed to load schema: %w", err)
+			}
+			for _, p := range plugins {
+				if inj, ok := p.(plugin.LateSourceInjector); ok {
+					if s := inj.InjectSourceLate(cfg.Schema); s != nil {
+						cfg.Sources = append(cfg.Sources, s)
+					}
+				}
+			}
+			// LoadSchema again now we have everything
```
Comment (a8m) [inline entgql/extension.go:147]:
```suggestion
			// LoadSchema again now we have everything.
```

File: entgql/internal/todofed/ent/entc.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	"entgo.io/contrib/entgql"
+	"entgo.io/ent/entc"
+	"entgo.io/ent/entc/gen"
+)
+
+func main() {
+	ex, err := entgql.NewExtension(
+		entgql.WithWhereFilters(true),
+		entgql.WithSchemaPath("../ent.graphql"),
+		entgql.WithConfigPath("../gqlgen.yml"),
```
Comment (a8m) [inline entgql/internal/todofed/ent/entc.go:22]: The example looks good, but don't you think we should add to this an example for passing [`api.Option`](https://github.com/99designs/gqlgen/blob/master/api/option.go#L8)s?

Comment (giautm) [reply entgql/internal/todofed/ent/entc.go:22]: There aren't have many options to demo. Built in options are NoPlugins and WithPlugIn that require some thirdpary implement. I think its complicated to demo.

File: entgql/extension.go [hunk 8]
```diff
@@ -90,7 +95,7 @@ const GQLConfigAnnotation = "GQLConfig"
 //
 // Note that, enabling this option is recommended as it improves the
 // GraphQL integration,
-func WithConfigPath(path string) ExtensionOption {
+func WithConfigPath(path string, gqlgenOpts ...api.Option) ExtensionOption {
```
[the remark this answers is not in the document]
Comment (giautm) [reply entgql/extension.go:98]: Maybe keeps it as gqlgenOptions? I want to make it clear that options come from the gqlgen.

[the remark this answers is not in the document]
Comment (a8m) [reply entgql/extension.go:98]: 👍

File: entgql/extension.go [hunk 9]
```diff
@@ -109,8 +114,39 @@ func WithConfigPath(path string) ExtensionOption {
...
+			if cfg.Model.IsDefined() {
+				plugins = append(plugins, modelgen.New())
+			}
+			plugins = append(plugins, resolvergen.New())
+			if cfg.Federation.IsDefined() {
+				plugins = append([]plugin.Plugin{federation.New()}, plugins...)
+			}
+			for _, o := range gqlgenOpts {
+				o(cfg, &plugins)
+			}
```
Comment (giautm) [inline entgql/extension.go:129]:
```suggestion
			for _, opt := range gqlgenOptions {
				opt(cfg, &plugins)
			}
```

File: entgql/extension.go (hunk 8 above)
Comment (giautm) [inline entgql/extension.go:98]:
```suggestion
func WithConfigPath(path string, gqlgenOptions ...api.Option) ExtensionOption {
```

File: entgql/internal/todofed/ent/entc.go [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+	"entgo.io/contrib/entgql"
+	"entgo.io/ent/entc"
+	"entgo.io/ent/entc/gen"
+)
+
+func main() {
+	ex, err := entgql.NewExtension(
+		entgql.WithWhereFilters(true),
+		entgql.WithSchemaPath("../ent.graphql"),
```
Comment (giautm) [inline entgql/internal/todofed/ent/entc.go:21]:
```suggestion
```

**Comment by a8m** [comment]
Thanks for the contribution @giautm ❤️

**Pull request merged**